File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        int n=arr.size();
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<>> pq;
        for(int i=0;i<n;i++)
        {
            pq.push({arr[i],i});
        }
        int rank=0,prevElement=INT_MIN;
        while(!pq.empty())
        {
            int index=pq.top().second;
            int currElement=pq.top().first;
            if(currElement==prevElement)
            {
                arr[index]=rank;
            }
            else
            {
                rank++;
                arr[index]=rank;
                prevElement=currElement;
            }
            pq.pop();
        }
        return arr;


        // unordered_map<int,int> m;
        // vector<int> result;
        // result.insert(result.end(),arr.begin(),arr.end());
        // sort(result.begin(),result.end());
        // int i,count=0;
        // int n=arr.size();
        // for(i=0;i<n;i++)
        // {
        //     if(m.find(result[i])==m.end())
        //     {
        //         count++;
        //         m[result[i]]=count;
        //     }
        // }
        // for(i=0;i<n;i++)
        // {
        //     arr[i]=m[arr[i]];
        // }
        // return arr;
    }
};

File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.java
class Solution {
    class Pair {
        int first;
        int second;

        public Pair(int first,int second) {
            this.first = first;
            this.second = second;
        }
    }

    public int[] arrayRankTransform(int[] arr) {
        int n = arr.length;
        PriorityQueue<Pair> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.first));

        for(int i=0;i<n;i++) {
            pq.offer(new Pair(arr[i],i));
        }

        int rank = 0, prev = Integer.MAX_VALUE;
        int[] result = new int[n];
        while(!pq.isEmpty()) {
            int index = pq.peek().second;
            pq.poll();
            if(arr[index] != prev) rank++;
            result[index] = rank;
            prev = arr[index];
        }
        return result;
    }
}